core/src/main/scala/org/apache/spark/executor/ProcfsMetricsGetter.scala

```diff
@@ -62,7 +62,21 @@ private[spark] class ProcfsMetricsGetter(procfsDir: String = "/proc/") extends L
         SparkEnv.get.conf.get(config.EVENT_LOG_STAGE_EXECUTOR_METRICS)
       val shouldLogStageExecutorProcessTreeMetrics =
         SparkEnv.get.conf.get(config.EVENT_LOG_PROCESS_TREE_METRICS)
-      procDirExists.get && shouldLogStageExecutorProcessTreeMetrics && shouldLogStageExecutorMetrics
+      val shouldAddProcessTreeMetricsToMetricsSet =
+        SparkEnv.get.conf.get(config.METRICS_PROCESS_TREE_METRICS)
+      val pickEitherUIOrMetricsSet = shouldLogStageExecutorProcessTreeMetrics ^
+        shouldAddProcessTreeMetricsToMetricsSet
+      val areBothUIMetricsEnabled = shouldLogStageExecutorProcessTreeMetrics &&
+        shouldAddProcessTreeMetricsToMetricsSet
+      if (areBothUIMetricsEnabled) {
+        logWarning("You have enabled " +
+          "both spark.eventLog.logStageExecutorProcessTreeMetrics.enabled" +
+          " and spark.metrics.logStageExecutorProcessTreeMetrics.enabled. This isn't " +
+          "allowed. As a result Procfs metrics won't be reported to UI or Metricsset")
+      }
+      (procDirExists.get && shouldLogStageExecutorMetrics && pickEitherUIOrMetricsSet) ||
```

Comment:
For process tree metrics that are going to be reported to REST API, logging for stage executor metrics should be enabled. Anyway using CachedGauge this may become simpler.

Reply:
I think if pickEitherUIOrMetricsSet is changed to use || operator, then the above return statement should suffice.

Regarding, caching of the metrics, I agree with that as well. I will just recommend to have caching built in ProcfsMetricsGetter.pTreeInfo.computeAllMetrics(). This method should just return last cached metrics if they are not stale, otherwise compute new metrics. This way it can be used by both of MetricSet and Logging.

Reply:
I also think that way is better, since it is simpler. However the pulling rate and heart beat aren't the same. So I guess we can document that to set expectations.